### ISSUE TITLE
Installing R.rsp causes HTML-based help.search() to give an error

### ISSUE THREAD

**Issue by HenrikBengtsson** [HenrikBengtsson/R.rsp#27]
From https://github.com/jeroen/jsonlite/issues/196:

# Issue

Having R.rsp installed can break HTML-based search to break.  For example, with:
```r
options(help_type = "html")
help.search("rsp")
```
may display
```
Error in if (nchar(Outfile)) Outfile else File : 
  argument is not interpretable as logical
```
in the browser.

# Troubleshooting

From some of the comments in https://github.com/jeroen/jsonlite/issues/196:

```r
debugonce(tools:::makeVignetteTable)
help.search("rsp")
[...]
Browse[2]> vignettes
     Package File Title PDF R 
[1,] "R.rsp" NA   NA    NA  NA
[2,] "R.rsp" NA   NA    NA  NA
[3,] "R.rsp" NA   NA    NA  NA
[4,] "R.rsp" NA   NA    NA  NA
[5,] "R.rsp" NA   NA    NA  NA
```

The problem appears to be that the `Topic` field below contains file extensions:
```r
> temp <- utils::help.search(pattern = "rsp", package = "R.rsp", type = "vignette")$matches
> temp$Topic
 [1] "Dynamic_document_creation_using_RSP.tex" "Dynamic_document_creation_using_RSP.tex"
 [3] "Dynamic_document_creation_using_RSP.tex" "RSP_intro.html"                         
 [5] "RSP_intro.html"                          "RSP_intro.html"                         
 [7] "RSP_refcard.tex"                         "RSP_refcard.tex"                        
 [9] "RSP_refcard.tex"                         "R_packages-RSP_vignettes.md"            
[11] "R_packages-RSP_vignettes.md"             "R_packages-RSP_vignettes.md"            
[13] "R_packages-Vignettes_prior_to_R300.tex" 
```

**Comment by HenrikBengtsson** [comment]
Further narrowing in on the problem:
```r
> db <- utils::hsearch_db("R.rsp", types = "vignette")
> temp <- subset(db$Base, Package == "R.rsp" & Type == "vignette")
> temp$Topic
[1] "R_packages-Static_PDF_and_HTML_vignettes.pdf" "R_packages-LaTeX_vignettes"                  
[3] "Dynamic_document_creation_using_RSP.tex"      "RSP_intro.html"                              
[5] "RSP_refcard.tex"                              "R_packages-RSP_vignettes.md"                 
[7] "R_packages-Vignettes_prior_to_R300.tex"      
```

Note how `Name` carry file-name extensions, which it shouldn't.  This [stems from an assumption in `utils` that vignette files only has one filename extension](https://github.com/wch/r-source/blob/118d3850054219eaab7ffe321349ceeeea88433d/src/library/utils/R/help.search.R#L46).  Also, [`Topic` <- `Name` by default](https://github.com/wch/r-source/blob/118d3850054219eaab7ffe321349ceeeea88433d/src/library/utils/R/help.search.R#L47), so the key problem is most likely that `Name` is invalid.

**Comment by kevinushey** [comment]
@HenrikBengtsson: do you think this is worth reporting to R-core?

**Comment by HenrikBengtsson** [comment]
We could. I'll try to do some more troubleshooting over the weekend to what's missing and if this is a recent bug or has always been there.  (When I said I'm sure I haven't seen it before, it could simply be that I've never done `help.search()` on topics where this happens).

**Comment by HenrikBengtsson** [comment]
To continue what I mentioned above, the [assumption that vignette files only has one filename extension](https://github.com/wch/r-source/blob/tags/R-3-4-1/src/library/utils/R/help.search.R#L46) was only true in R (< 3.0.0).  In R (>= 3.0.0), it's possible for vignette engines to specify an arbitrary filename pattern.  For instance, engines `Sweave`, `knitr::knitr`, and `R.rsp::tex` use:
```r
 tools::vignetteEngine("Sweave")$pattern
[1] "[.][rRsS](nw|tex)$"

> tools::vignetteEngine("knitr::knitr")$pattern
[1] "[.]([rRsS](nw|tex)|[Rr](md|html|rst))$"

> tools::vignetteEngine("R.rsp::tex")$pattern
[1] "[.](tex|ltx)$"
```
These meet the assumption of a single level of filename extension such that `sub("\\.[^.]*$", "", basename(vDB$File))` suffice.

However, engines such as `R.rsp::rsp` and `R.rsp::asis` use filenames with two levels of extensions:
```r
> tools::vignetteEngine("R.rsp::rsp")$pattern
[1] "[.][^.]*[.]rsp$"

> tools::vignetteEngine("R.rsp::asis")$pattern
[1] "[.](pdf|html)[.]asis$"
```

A proper fix to `utils:::merge_vignette_index()` would involve replacing:
```r
base[, "Name"] <- sub("\\.[^.]*$", "", basename(vDB$File))
```
with something like:
```r
engine_pattern <- tools::vignetteEngine(engine_name)$pattern
base[, "Name"] <- sub(engine_pattern, "", basename(vDB$File))
```

The problem with this "at-runtime" approach is that the following have to happen each time the vignette database is (re-)built:

1. the vignette engine name needs to be inferred
2. the vignette engine package (`VignetteBuilder`) needs to be loaded

See `tools::pkgVignettes()` for how this can done (although inefficiently).


However, a more efficient solution would probably be to set the vignette `Name` when the package vignette database is built, i.e. during installation, which occurs in [`tools:::.build_vignette_index()`](https://github.com/wch/r-source/blob/tags/R-3-4-1/src/library/tools/R/Vignettes.R#L804-L869) called by [`tools:::.install_package_vignettes2()`](https://github.com/wch/r-source/blob/tags/R-3-4-1/src/library/tools/R/admin.R#L471-L586).

Yet another solution would be to record the vignette-engine pattern (also in `tools:::.build_vignette_index()`) for each vignette such that it's available in `vDB`;
```r
base[, "Name"] <- sub(vDB$Pattern, "", basename(vDB$File))
```

**Comment by HenrikBengtsson** [comment]
A third, and much simpler, alternative could be to replace:
```r
base[, "Name"] <- sub("\\.[^.]*$", "", basename(vDB$File))
```
with
```r
base[, "Name"] <- sub("\\.[^.]*$", "", basename(vDB$PDF))
```
This would move the assumption of a single-level of filename extensions on the vignette _source file_ to the same assumption on the vignette _output file_, which (currently) may only be `*.PDF` and `*.HTML`.

**Comment by HenrikBengtsson** [comment]
Ah!  The original error originates from [`tools:::httpd()`](https://github.com/wch/r-source/blob/871172dfa2798ea401be4982c85a4653a0144de8/src/library/tools/R/dynamicHelp.R#L136-L143):
```r
topic <- temp[i, "Topic"]
[...]
vignette <- vigDB[topic == file_path_sans_ext(vigDB$PDF),]
```
which suggests that [`utils:::merge_vignette_index()`](https://github.com/wch/r-source/blob/tags/R-3-4-1/src/library/utils/R/help.search.R#L46) should indeed use:
```r
base[, "Name"] <- sub("\\.[^.]*$", "", basename(vDB$PDF))
base[, "Topic"] <- base[, "Name"]
```
rather than
```r
base[, "Name"] <- sub("\\.[^.]*$", "", basename(vDB$File))
base[, "Topic"] <- base[, "Name"]
```

This change / bug fix seems easier to argue for.

PS. The conclusions in this last comment looks awfully familiar to me - I might have discovered this bug in the past for other reasons - it could also just  be a jet lag-induced deja vu.

**Comment by HenrikBengtsson** [comment]
@kevinushey, does the conclusions in https://github.com/HenrikBengtsson/R.rsp/issues/27#issuecomment-325071451 about a bug in `utils:::merge_vignette_index()` look reasonable to you?

**Comment by kevinushey** [comment]
Nice sleuthing! It sounds to me like the bug lives there as well.

**Comment by HenrikBengtsson** [comment]
I just sent a bug report with below patch R-devel (https://stat.ethz.ch/pipermail/r-devel/2017-August/074834.html).
```diff
Index: src/library/utils/R/help.search.R
===================================================================
--- src/library/utils/R/help.search.R (revision 73159)
+++ src/library/utils/R/help.search.R (working copy)
@@ -43,7 +43,7 @@
  base[, "LibPath"] <- path
  id <- as.character(1:nrow(vDB) + NROW(hDB[[1L]]))
  base[, "ID"] <- id
- base[, "Name"] <- sub("\\.[^.]*$", "", basename(vDB$File))
+ base[, "Name"] <- tools::file_path_sans_ext(basename(vDB$PDF))
  base[, "Topic"] <- base[, "Name"]
  base[, "Title"] <- vDB$Title
  base[, "Type"] <- "vignette"
```

**Comment by HenrikBengtsson** [comment]
UPDATE: Fixed in [R-devel (rev 73170)](https://github.com/wch/r-source/commit/be85b05ff2f9f4149d328c0923f8f4b21837bb14).

**Comment by HenrikBengtsson** [comment]
UPDATE: Fix ported to the [R 3.4.x branch (rev 73176)](https://github.com/wch/r-source/commit/df43b4de0285ee02db4176104a04e514041b768d), meaning it should be part of the R 3.4.2 release and I guess also R 3.4.1 patched.
